compiler/luci/pass/src/FuseMulWithConvPass.cpp

```diff
@@ -0,0 +1,209 @@
+ *
+ *  BEFORE
+ *                    |
+ *              [CircleConv2D] (no activation)
+ *                    |
+ *                  [Mul] (channel-wise/scalar constant)
+ *                    |
+ *
+ *  AFTER
+ *                    |
+ *            [CircleConv2D] (with updated kernels, bias, and activation)
+ *                    |
+ *
+ */
+
+bool fuse_mul_with_conv(luci::CircleMul *mul)
+{
+  // sanity check
+  RETURN_FALSE_UNLESS(mul->dtype() == loco::DataType::FLOAT32);
+
+  luci::CircleConst *const_mul_operand = nullptr;
+  luci::CircleConv2D *conv = nullptr;
+  RETURN_FALSE_UNLESS(luci::fill(&const_mul_operand, &conv).with_commutative_args_of(mul));
+
+  // sanity check
+  RETURN_FALSE_UNLESS(conv->dtype() == loco::DataType::FLOAT32 &&
+                      const_mul_operand->dtype() == loco::DataType::FLOAT32);
+
+  //  NOTE for general activation function: S * Act(A * B) != Act(A*(SB))
+  if (conv->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return false;
+
+  // check that const_mul_operand is channel-wise constant or just a scalar
+  RETURN_FALSE_UNLESS(const_mul_operand->rank() == 4 || const_mul_operand->rank() == 1 ||
+                      const_mul_operand->rank() == 0);
+
+  std::vector<float> mul_values;
+  if (const_mul_operand->rank() == 4)
+  {
+    // check channel-wise broadcasting
+    RETURN_FALSE_UNLESS(const_mul_operand->dim(0).value() == 1 &&
+                        const_mul_operand->dim(1).value() == 1 &&
+                        const_mul_operand->dim(2).value() == 1);
+  }
+  else if (const_mul_operand->rank() == 1 || const_mul_operand->rank() == 0)
+  {
+    // sanity check
+    RETURN_FALSE_UNLESS(const_mul_operand->size<loco::DataType::FLOAT32>() != 0);
+  }
+
+  mul_values.resize(const_mul_operand->size<loco::DataType::FLOAT32>());
+  for (uint32_t idx = 0; idx < mul_values.size(); idx++)
+  {
+    mul_values[idx] = const_mul_operand->at<loco::DataType::FLOAT32>(idx);
+  }
+
+  // filter
+  auto const conv_filter = dynamic_cast<luci::CircleConst *>(conv->filter());
+
+  // sanity check
+  RETURN_FALSE_UNLESS(conv_filter != nullptr && conv_filter->rank() == 4 &&
+                      conv_filter->dtype() == loco::DataType::FLOAT32);
+
+  auto const out_channels = conv_filter->dim(0).value();
+
+  // multiplier is either channelwise constant or scalar
+  RETURN_FALSE_UNLESS(out_channels == mul_values.size() || mul_values.size() == 1);
+
+  // bias
+  auto const conv_bias = dynamic_cast<luci::CircleConst *>(conv->bias());
+
+  RETURN_FALSE_UNLESS(conv_bias == nullptr ||
+                      (conv_bias->rank() == 1 && conv_bias->dim(0) == out_channels &&
+                       conv_bias->dtype() == loco::DataType::FLOAT32));
+
+  luci::CircleConst *fused_conv_filter = nullptr;
+  {
+    // fused filter
+    fused_conv_filter = luci::clone(conv_filter);
+    assert(mul_values.size() == 1 or mul_values.size() == out_channels); // FIX_ME_UNLESS
```

Comment:
1/ let's use `||` or `or` for consistency.
2/ why is `assert` used instead of `RETURN_FALSE_UNLESS` ?

Reply:
@jinevening @seanshpark  
Please see Ln 116
```
RETURN_FALSE_UNLESS(out_channels == mul_values.size() || mul_values.size() == 1);
```
May be it would be enough (no need for additional asserts)?

Reply:
looks duplicate.

Reply:
@jinevening 
Which one should be left? Or both?

Reply:
Ah, the assert can be removed. I missed it. Sorry for confusion.

Reply:
ok. Thank you.

Reply:
removed.